### PR TITLE
Player Trading and Item Transfer Flags

### PR DESCRIPTION
**Pull request by Zintixx**
- Implements player trading
- Implements transfer flags (Thanks @kOchirasu for the help)
 - Still no idea what 32 bit flag is but it doesn't look like it's in use?
- Adjusted item packet to display/behave correct details about item repackaging, tradability, and trade count.
- Fixed a couple of misc missing packets and some debugging lines.

### CODE REVIEW

File: MapleServer2/PacketHandlers/Game/TradeHandler.cs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        if (tradeInventory is null)
+        {
+            return;
+        }
+
+        Player otherPlayer = tradeInventory.OtherPlayer;
+
+        // Return items back to player's inventory
+        session.Player.TradeInventory.SendItems(session.Player, false);
+        session.Player.TradeInventory = null;
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/TradeHandler.cs:124]:
```suggestion
        tradeInventory.SendItems(session.Player, false);
        tradeInventory = null;
```

File: MapleServer2/PacketHandlers/Game/TradeHandler.cs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        {
+            item.TrySplit(amount, out Item splitItem);
+            session.Send(ItemInventoryPacket.UpdateAmount(itemUid, item.Amount));
+            item = splitItem;
+        }
+        else
+        {
+            session.Player.Inventory.RemoveItem(session, itemUid, out Item removedItem);
+            item = removedItem;
+        }
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/TradeHandler.cs:163]: Can't you use **Inventory.Consume** here?

File: MapleServer2/PacketHandlers/Game/TradeHandler.cs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            return;
+        }
+
+        Player player = session.Player.TradeInventory.OtherPlayer;
+        session.Player.TradeInventory.AlterTrade(session);
+
+        session.Player.TradeInventory.Mesos = mesoAmount;
+
+        session.Send(TradePacket.AddMesosToTrade(mesoAmount, true));
+        player.Session?.Send(TradePacket.AddMesosToTrade(mesoAmount, false));
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/TradeHandler.cs:196]:
```suggestion
        Player otherPlayer = session.Player.TradeInventory.OtherPlayer;
        session.Player.TradeInventory.AlterTrade(session);

        session.Player.TradeInventory.Mesos = mesoAmount;

        session.Send(TradePacket.AddMesosToTrade(mesoAmount, true));
        otherPlayer.Session?.Send(TradePacket.AddMesosToTrade(mesoAmount, false));
```

File: MapleServer2/Types/Inventory.cs [hunk 4]
```diff
@@ -169,8 +180,11 @@ public void AddItem(GameSession session, Item item, bool isNew)
 
                     DatabaseManager.Items.Delete(item.Uid);
 
-                    session.Send(ItemInventoryPacket.Update(existingItem.Uid, existingItem.Amount));
-                    session.Send(ItemInventoryPacket.MarkItemNew(existingItem, item.Amount));
+                    session.Send(ItemInventoryPacket.UpdateAmount(existingItem.Uid, existingItem.Amount));
+                    if (notify)
+                    {
+                        session.Send(ItemInventoryPacket.MarkItemNew(existingItem, item.Amount));
+                    }
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/Inventory.cs:187]: Isn't **notify** the exact same thing as **isNew**??

File: MapleServer2/Types/TradeInventory.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public void AlterTrade(GameSession session)
+    {
+        if (OtherPlayer.TradeInventory.IsLocked)
+        {
+            IsLocked = false;
+            OtherPlayer.Session?.Send(TradePacket.OfferedAltered(true));
+            session.Send(TradePacket.OfferedAltered(false));
+        }
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/TradeInventory.cs:42]: invert if

File: MapleServer2/Commands/Game/ItemCommands.cs [hunk 6]
```diff
@@ -37,6 +37,7 @@ public override void Execute(GameCommandTrigger trigger)
         Item item = new(itemId)
         {
             Rarity = rarity >= 0 ? rarity : ItemMetadataStorage.GetRarity(itemId),
+            TransferFlag = ItemMetadataStorage.GetTransferFlag(itemId, rarity),
```
Comment (AngeloTadeucci) [inline MapleServer2/Commands/Game/ItemCommands.cs:40]: Doesn't this already happens with **SetMetadataValues** on item creation?

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/Commands/Game/ItemCommands.cs (hunk 6 above)
Comment (Zintixx) [reply MapleServer2/Commands/Game/ItemCommands.cs:40]: The command is technically scuffed last time we talked about it.
The rarity is not displayed properly until after you change maps/relog. (try spawning an item and chat linking it)
And with rarity not correctly updating till then, it also affects the transfer flags for this.

File: MapleServer2/PacketHandlers/Game/TradeHandler.cs (hunk 2 above)
Comment (Zintixx) [reply MapleServer2/PacketHandlers/Game/TradeHandler.cs:163]: No because we're not deleting the item. I also need to extract the item out of said stack.

**Review by AlanMorel — APPROVED** [review]

**Review by AngeloTadeucci — APPROVED** [review]

**Review by darkvergus — APPROVED** [review]

**Pull request merged**